Pandas DataFrame/Networkx graph to HeteroData

ramshi236 opened the pull request:
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203610797289918
  - https://app.asana.com/0/0/1202443136355917

ramshi236 committed:
new modules and tests

ramshi236 committed:
changlog update and setup.py

ramshi236 committed:
little fix

ramshi236 committed:
fixing format in code

ramshi236 committed:
docstring fixes

ramshi236 committed:
fixes

ramshi236 committed:
fixes

ramshi236 commented:
although i pass it locally, it seems that i can't pass the remote pre-commit-ci. 
do you have any idea why? @rusty1s

rusty1s commented:
Looks like `pre-commit` cannot push to your PyG fork - this looks to be a permission issue.

Commit:
rerun yapf manually on new files

ramshi236 committed:
Merge branch 'master' into bt_HeteroDF

ramshi236 commented:
@rusty1s can you review ?

ramshi236 committed:
Merge branch 'master' into bt_HeteroDF

rusty1s commented:
Just got back from vacation - will do a review soon.

rusty1s commented on torch_geometric/datasets/pandas_heterodata/hetero_df.py:
```diff
@@ -0,0 +1,343 @@
+    Example:
+        .. code-block::
+
+            college_nx = nx.MultiDiGraph() college_nx.add_nodes_from( [ (1,
+            {"id": 1, "name": "bob", "node_type": "student", }), (2, {"id":
+            2, "name": "dan", "node_type": "student", }), (3, {"id": 3,
+            "name": "tal", "node_type": "ta", }), (4, {"id": 4, "name":
+            "profG", "node_type": "lecturer", }), (5, {"id": 5, "name":
+            "phys101", "node_type": "class", }), (6, {"id": 6, "name":
+            "phys102", "node_type": "class", }), (7, {"id": 7, "name":
+            "chem101", "node_type": "class", }), (8, {"id": 8, "name":
+            "physics", "node_type": "topic"}), (9, {"id": 9, "name":
+            "chemistry", "node_type": "topic"}),
+            ] ) college_nx.add_edges_from( [ (1, 5, {"edge_type": "takes"}),
+            (1, 7, {"edge_type": "takes"}), (2, 6, {"edge_type": "takes"}),
+            (3, 5, {"edge_type": "teaches"}), (4, 6, {"edge_type":
+            "teaches"}), (5, 8, {"edge_type": "is_about"}), (6, 8,
+            {"edge_type": "is_about"}), (6, 9, {"edge_type": "is_about"}),
+            (7, 9, {"edge_type": "is_about"}), ] ) nxHeteroDF =
+            nx_to_HeteroDF(college_nx) print(nxHeteroDF.keys()) # >>>
+            dict_keys(['class', 'lecturer', 'student', 'ta', 'topic',
+            ('class', 'is_about', 'topic'), # ('lecturer', 'teaches',
+            'class'), ('student', 'takes', 'class'), ('ta', 'teaches',
+            'class')])
```
Can we fix the formatting here?

rusty1s commented on torch_geometric/datasets/pandas_heterodata/hetero_df.py:
```diff
@@ -0,0 +1,343 @@
+    except with "source_type" and "target_type" columns derived from the
+    nodes df a signature df which is like the edge df except with
+    "source_type" and "target_type" columns derived from the nodes df ..
+    code-block:: nodes = pd.DataFrame({"id": [1,2,3,4,5], "node_type": ["a",
+    "a","a","b","b"]}) edges = pd.DataFrame({"source": [1,2,3,5], "target":[
+    2,4,5,1], "edge_type": ["c","c","c","c"]}) print(
+    node_edge_dfs_to_sig_df(nodes, edges)) source  target source_type
+    edge_type target_type 0       1       2           a         c
+    a 1       2       4           a         c           b 2       3       5
+             a         c           b 3       5       1           b         c
+                       a
```
Can we fix the formatting here?

rusty1s commented on torch_geometric/datasets/pandas_heterodata/hetero_df.py:
```diff
@@ -0,0 +1,343 @@
+        -> Dict[Any, pd.DataFrame]:
+    r"""groups a dataframe by a col or list of cols and returns a dictionary
+    of {groupby_values:group_df}, return a dictionary with the group by
+    values as keys and the grouped dfs as values. .. code-block:: df =
+    pd.DataFrame({ "student": ["Ram", "Noa", "Song", "Ram", "Ram", "Song",
+    "Noa", "Noa"], "grades" : [100, 80, 90, 55, 50, 60, 70, 80], "class" : [
+    "history","history","history", "math", "art","history", "math", "art"]
+    }) print(groupby_to_dict(df, "student")) {'Noa':    grades    class 1
+      80  history 6      70     math 7      80      art, 'Ram':    grades
+      class 0     100  history 3      55     math 4      50      art,
+      'Song':    grades    class 2      90  history 5      60  history}
+    """
```
Can we fix the formatting here?

rusty1s commented on torch_geometric/datasets/pandas_heterodata/__init__.py:
```diff
@@ -0,0 +1,10 @@
+from .hetero_df import (nx_to_HeteroDF, heteroDF_split, heteroDF_merge,
```
Is there a need to create a new sub-package?

ramshi236 replied:
i think so because we introduced a new data structure with from/to conversion ,and the split and merge functions are working only with that new data structure (`HetreoDF`)

rusty1s commented on torch_geometric/datasets/pandas_heterodata/hetero_df.py:
```diff
@@ -0,0 +1,343 @@
+    targetType) values are pd.Dataframes with source and target columns that
+    contain the node ids that define the edge and other feature columns.
+
+    Args: hdf: :class:`~torch_geometric.typing.HeteroDF`. ignore: whether to
+    ignore some features of the heteroDF. takes either a list of columns to
+    ignore for each dataframe or a heteroColumnSet. keep: used only if
+    ignore is None. whether to keep some features of the heteroDF. takes
+    either a list of columns to keep for each dataframe or a
+    :class:`~torch_geometric.typing.HeteroColumnSet`.
```
Can we fix formatting here?

rusty1s commented on torch_geometric/datasets/pandas_heterodata/hetero_df.py:
```diff
@@ -0,0 +1,343 @@
+    we want a dataframe with node_id/source target pairs to still exist in
+    the not matched/matched part.
+
+    Args: hdf (:class:`~torch_geometric.typing.HeteroDF`): HeteroDf to
+    split. split_by (:obj:`Union[List[str]]`,
+    :class:`~torch_geometric.typing.HeteroColumnSet`): keep_index_in_matched
+    (:obj:`bool`): keeping index in matched HeteroDF keep_index_in_unmatched
+    (:obj:`bool`): keeping index in unmatched HeteroDF
```
Can we fix formatting here?

rusty1s reviewed (changes requested):
Thanks for the PR. Initially I thought of moving this to `dataset` to provide a `HeteroDFDataset(InMemoryDataset)` class that takes in a list of `HeteroDF` objects and convert them to `HeteroData` objects. The current implementation looks like it is better placed in `utils`.

I am also wondering why we don't start more simple and add a homogeneous graph converter first. A simple function that takes in a node `DataFrame` and an edge `DataFrame` that gets combined into a `Data` object. I think this implementation would ease reviewing and would help us to make progress towards this PR more quickly. WDYT? I would also prefer to move `pandas`/`networkx` converters into different PRs.

ramshi236 committed:
merge with master

ramshi236 committed:
moved tests

ramshi236 committed:
typos

ramshi236 committed:
Merge branch 'master' into bt_HeteroDF

ramshi236 committed:
changelog update

ramshi236 committed:
fix for tests

ramshi236 committed:
docstriong typos

ramshi236 committed:
docstring fix

ramshi236 committed:
docstring fix

ramshi236 committed:
fix docstrings

ramshi236 commented:
I've addressed your formatting issues and moved hetero_df.py to torch_geometric/utils.
I've also added one more function `heteroData2heteroDf`.
i don't think this is a good idea to split the new logic (ie hetero_df.py) to a networkx and pandas pr, because all the functions there are transformations from/to `HeteroDF` object.

about your suggestion for a simple homogeneous converter example, you can do it simply by :
```python
nodes_df={
'A':pd.DataFrame.from_records([
    {"id": 0, "feat1":1.0,"feat2":2.0,},
    {"id": 1, "feat1":11.0,"feat2":12.0,}
])}
edges_df ={
('A','to','A'):pd.DataFrame.from_records([
    {"source": 0,"target":1}
])}

homo_df:HeteroDF = {**nodes_df,**edges_df} 
# HeteroDF is just an alias to a Dict[str|Tuple[str,str,str],DataFrame] (so the name HeteroDF can misguide)
hetero_data,_ =heteroDf_to_HeteroData(homo_df)
hetero_data.to_homogeneous().to_heterogeneous()
# HeteroData(
#   A={ x=[2, 2] },
#   (A, to, A)={ edge_index=[2, 1] }
# )
```

ramshi236 committed:
Merge branch 'master' into bt_HeteroDF
